File: HomeWork1/Hw1/Parser.cs


namespace Hw1;

public static class Parser
{
    public static void ParseCalcArguments(string[] args, 
        out double val1, 
        out CalculatorOperation operation, 
        out double val2)
    {
        if (!IsArgLengthSupported(args))
            throw new ArgumentException("Incorrect length:", nameof(args));
        if (!Double.TryParse(args[0], out val1))
            throw new ArgumentException("Incorrect first value:", nameof(val1));
        if (!Double.TryParse(args[2], out val2))
            throw new ArgumentException("Incorrect second value:", nameof(val2));
        operation = ParseOperation(args[1]);
    }

    private static bool IsArgLengthSupported(string[] args) => args.Length == 3;

    private static CalculatorOperation ParseOperation(string arg)
    {
        return arg switch
        {
            "+" => CalculatorOperation.Plus,
            "-" => CalculatorOperation.Minus,
            "*" => CalculatorOperation.Multiply, 
            "/" => CalculatorOperation.Divide,
            _=> throw new InvalidOperationException()
        };
    }
}

File: Homework4/Hw4/Parser.fs
﻿module Hw4.Parser

open System
open Hw4.Calculator
open Microsoft.FSharp.Core


type CalcOptions = {
    arg1: float
    arg2: float
    operation: CalculatorOperation
}

let isArgLengthSupported (args : string[]) =
    match Array.length args with
    | 3 -> true
    | _ -> false

let parseOperation (arg : string) =
    match arg with
    | "+" -> CalculatorOperation.Plus
    | "-" -> CalculatorOperation.Minus
    | "*" -> CalculatorOperation.Multiply
    | "/" -> CalculatorOperation.Divide
    | _ -> CalculatorOperation.Undefined
    
let tryParseValues (arg1 : string, arg2 : string) =
    match Double.TryParse arg1, Double.TryParse arg2 with
    | (true,_),(true,_) -> true
    | _ -> false
    
let parseCalcArguments(args : string[]) =
    match isArgLengthSupported args with
    | false -> ArgumentException("There must be 3 args") |> raise
    | true ->
        let operation = parseOperation args[1]
        match operation with
        | CalculatorOperation.Undefined -> ArgumentException("Undefined operation") |> raise
        | _ ->
            match tryParseValues (args[0], args[2]) with
            | true -> {arg1 = Double.Parse args[0]; arg2 = Double.Parse args[2]; operation = operation}
            | false -> ArgumentException("First or second value is not double") |> raise